### PR TITLE
Add header support

### PR DESCRIPTION
**Pull request by isaachess**
Add the ability to set arbitrary tag-name and value pairs when writing
to the buffer. Each tag and value is written to the header of the
buffer.

### CODE REVIEW

File: admproto/reader.go [hunk 1]
```diff
@@ -39,33 +39,68 @@ func (r *Reader) Begin(in []byte) (out, application, instance_id []byte, err err
...
-		return nil, nil, nil, Error.New("unknown version: %d", version[0])
+		return nil, nil, nil, 0, Error.New("unknown version: %d", version[0])
 	}
 
 	// determine if headers are included from the version
+	var hasTags bool
 	switch version[0] & headerMask {
 	case headersExcluded:
 	case headersIncluded:
+		hasTags = true
```
Comment (zeebo) [inline admproto/reader.go]: Can we call it `hasHeaders`?

File: admproto/reader.go [hunk 2]
```diff
@@ -24,10 +24,10 @@ func (r *Reader) Reset() {
 
 // Begin returns the header information out of the packet, and the remaining
 // data in the packet.
-func (r *Reader) Begin(in []byte) (out, application, instance_id []byte, err error) {
+func (r *Reader) Begin(in []byte) (out, application, instance_id []byte, numTags int, err error) {
```
Comment (zeebo) [inline admproto/reader.go]: Can we use snake_case for consistency?

Comment (isaachess) [reply admproto/reader.go]: My linter will continue to loudly complain, but yes.

File: admproto/proto_test.go [hunk 3]
```diff
@@ -31,12 +38,28 @@ func TestReaderWriter(t *testing.T) {
...
+		var headers map[string]string
+		if num_headers > 0 {
+			headers = make(map[string]string, num_headers)
+		}
+		var header_key, header_value []byte
+		for i := 0; i < num_headers; i++ {
+			buf, header_key, header_value, err = r.NextHeader(buf)
+			if err != nil {
+				t.Fatal("err reading key", err)
+			}
```
Comment (zeebo) [inline admproto/proto_test.go]:
```suggestion
			assertNoError(t, err)
```

File: admmonkit/monkit.go [hunk 4]
```diff
@@ -70,10 +73,13 @@ func Send(ctx context.Context, opts Options) (err error) {
 			// always ensure the buffer has the prefix in it.
 			if len(buf) == 0 {
 				// if we can't add the application and instance id, it's fatal.
-				buf, err = w.Begin(buf, opts.Application, opts.InstanceId)
+				buf, err = w.Begin(buf, opts.Application, opts.InstanceId, len(opts.Headers))
 				if err != nil {
 					return
 				}
+				for key, value := range opts.Headers {
+					buf, err = w.AppendHeader(buf, []byte(key), []byte(value))
```
Comment (zeebo) [inline admmonkit/monkit.go]:
```suggestion
					buf, err = w.AppendHeader(buf, []byte(key), []byte(value))
					if err != nil {
						return
					}
```

**Pull request merged**